content/00.build/65.developer-reference/50.era-contracts/60.elliptic-curve-precompiles.md

```diff
@@ -0,0 +1,311 @@
+The modulus is less than 256 bits, which is why every element in the field is represented as a `uint256`.
+
+The arithmetic is carried out with the field elements encoded in the Montgomery form.
+This is done not only because operating in the Montgomery form speeds up the computation but also because the native modular multiplication,
+which is carried out by Yul's `mulmod` opcode, is very inefficient.
+
+Instructions set on ZKsync and EVM are different, so the performance of the same Yul/Solidity code can be efficient on EVM, but not on zkEVM and opposite.
+
+One such very inefficient command is `mulmod`.
+On EVM there is a native opcode that makes modulo multiplication and it costs only 8 gas,
+which compared to the other opcodes costs is only 2-3 times more expensive.
+On zkEVM we don’t have native `mulmod` opcode, instead, the compiler does full-with multiplication
+(e.g. it multiplies two `uint256`s and gets as a result an `uint512`).
+Then the compiler performs long division for reduction (but only the remainder is kept),
+in the generic form it is an expensive operation and costs many opcode executions,
+which can’t be compared to the cost of one opcode execution.
+The worst thing is that `mulmod` is used a lot for the modulo inversion, so optimizing this one opcode gives a huge benefit to the precompiles.
+
+### Multiplication
+
+As said before, multiplication was carried out by implementing the Montgomery reduction,
+which works with general moduli and provides a significant speedup compared to the naïve approach.
+
+The squaring operation is obtained by multiplying a number by itself.
+However, this operation can have an additional speedup by implementing the SOS Montgomery squaring.
+
+### Inversion
+
+Inversion was performed using the extended binary Euclidean algorithm (also known as extended binary greatest common divisor).
+This algorithm is a modification of Algorithm 3 `MontInvbEEA` from [Montgomery inversion](https://cetinkayakoc.net/docs/j82.pdf).
+
+### Exponentiation
+
+The exponentiation was carried out using the square and multiply algorithm, which is a standard technique for this operation.
+
+## Montgomery Form
+
+Let’s take a number `R`, such that `gcd(N, R) == 1` and `R` is a number by which we can efficiently divide and take module over it
+(for example power of two or better machine word, aka 2^256).
+Then transform every number to the form of `x * R mod N` / `y * R mod N` and then we get efficient modulo addition and multiplication.
+The only thing is that before working with numbers we need to transform them to the form from `x mod N`  to the `x * R mod N`
+and after performing operations transform the form back.
+
+For the latter, we will assume that `N` is the module that we use in computations,
+and `R` is $2^{256}$, since we can efficiently divide and take module over this number
+and it practically satisfies the property of `gcd(N, R) == 1`.
```

Comment:
We can correct the instances of `module` here:
```suggestion
For the latter, we will assume that `N` is the modulus that we use in computations,
and `R` is $2^{256}$, since we can efficiently divide and take modulo over this number
and it practically satisfies the property of `gcd(N, R) == 1`.
```